### ISSUE TITLE
JDK11 Support

### ISSUE THREAD

**Issue by jdonenine** [k8ssandra/k8ssandra#979]
Content originally outlined by @adejanovski in #484 

## JDK11 support with Shenandoah & ZGC
[Benchmarks](https://thelastpickle.com/blog/2020/06/29/cassandra_4-0_garbage_collectors_performance_benchmarks.html) comparing 3.11 vs 4.0~alpha4 showed tremendous improvement in throughput and especially latencies (-88% at p99 when using Shenandoah).

Shenandoah is available in JDK11 using specific builds:

```
Releases
Shenandoah availability differs by vendor and JDK release. OpenJDK 12+ builds normally include Shenandoah by default. OpenJDK 11 requires the opt-in during build time.

Known vendor status is:

Red Hat
Fedora 24+ OpenJDK 8+ builds include Shenandoah

RHEL 7.4+ ships with OpenJDK 8+ that includes Shenandoah as Technology Preview
Red Hat OpenJDK 8u builds for Windows include Shenandoah
Amazon
Ships Shenandoah in Amazon Corretto, starting with OpenJDK 11.0.9
Oracle
Does not ship Shenandoah in any release, both OpenJDK builds and proprietary builds
Azul
Ships Shenandoah in Azul Zulu, starting with OpenJDK 11.0.9
SAP
(status unknown)
AdoptOpenJDK
Ships Shenandoah in default binaries, starting with OpenJDK 11.0.9
Linux distributions
Debian ships Shenandoah starting with OpenJDK 11.0.9
Gentoo ebuilds for IcedTea have Shenandoah USE flag
RHEL/Fedora-based distros or other distros that use packages from them may also have Shenandoah enabled. Notably, CentOS, Oracle Linux, and Amazon Linux are known to ship it.
```

I assume this would require us to create custom Apache Cassandra docker images to use with alternate JDKs.
Performance wise this is something that should be seriously considered though.

**Comment by Miles-Garnsey** [comment]
RE 4.0.0 features: I'm working on audit + query logging and client backpressure - these are close to complete (client backpressure is actually just awaiting review).

This particular ticket is going to be more in-depth however. I we should probably segregate the management API and Cassandra images before we move on this. 

Otherwise we'll need to build custom management API images for each of the JDK/OS versions we plan to support, which seems like a lot of surface area to test and maintain. @jdonenine and @jsanda what are your thoughts on chasing this particular rabbit?

**Comment by jsanda** [comment]
I agree about it being a lot of surface area. A few of questions....

* What JDK/OS versions do we want to support?
* Would it make sense to track each JDK/OS version we plan to support in its own ticket?
* Do we need to be concerned about the different OS versions? Is it sufficient to pick one and use that?

**Comment by Miles-Garnsey** [comment]
I incline towards CentOS but some people are [unhappy](https://www.theregister.com/2021/01/26/killing_centos/) with their new release model. If we wanted to indulge that view we might consider a custom build using Alpine which may confer an additional advantage w.r.t image size. 

It may be sufficient to pick one distro, but I think we should also support at least one non-Shenandoah image as well (given JDK 11 is still experimental in the upstream Cassandra project), so we're still expanding surface area if we don't re-imagine some other aspect of our image build process.

**Comment by jsanda** [comment]
@bradfordcp what are your thoughts especially taking UBI into consideration?

**Comment by bradfordcp** [comment]
My preference would be OpenJDK 8 & 11 on `redhat/ubi8-minimal`.

**Comment by bradfordcp** [comment]
Hey team! Please [add your planning poker estimate](https://app.zenhub.com/workspaces/K8ssandra-Grooming-623b197f1fcd4800110540ef/issues/k8ssandra/k8ssandra/979?planning-poker) with ZenHub @Miles-Garnsey @jsanda

**Comment by jsanda** [comment]
These changes should be considered for k8ssandra-operator as should planning poker estimates.

**Comment by Miles-Garnsey** [comment]
I recommend evaluating whether changes should land in cass-operator initially (to ensure that DSE benefits - if possible) and be surfaced via k8ssandra-operator.

This ticket may become complex if config builder is involved and end up spreading out over 3-4 repos (management API images, config builder, k8ssandra-operator, cass-operator).

I would approach this with caution.

**Comment by adutra** [comment]
Related: https://github.com/k8ssandra/k8ssandra-operator/issues/490.

I'd also note that currently the mgmt-api images we ship have only the following 2 combinations:

* Cassandra 3.11 + OpenJDK 8
* Cassandra 4.0 + OpenJDK 11
